driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1074,15 +1085,14 @@ private PooledConnection acquirePermitOrGetAvailableOpenedConnection(final boole
                     expressDesireToGetAvailableConnection();
                     expressedDesireToGetAvailableConnection = true;
                 }
-                long remainingNanos = timeout.remainingOrInfinite(NANOSECONDS);
                 while (permits == 0
                         // the absence of short-circuiting is of importance
                         & !stateAndGeneration.throwIfClosedOrPaused()
                         & (availableConnection = tryGetAvailable ? tryGetAvailableConnection() : null) == null) {
-                    if (Timeout.expired(remainingNanos)) {
-                        throw createTimeoutException(timeout);
+                    if (timeout.hasExpired()) {
+                        throw createTimeoutException(startTime);
                     }
-                    remainingNanos = awaitNanos(permitAvailableOrHandedOverOrClosedOrPausedCondition, remainingNanos);
+                    timeout.awaitOn(permitAvailableOrHandedOverOrClosedOrPausedCondition);
                 }
```

Comment:
Note that the timeout is now checked directly, rather than the value from await.

Reply:
Ack